### ISSUE TITLE
`unnecessary_safety_comment` and `undocumented_unsafe_blocks` visit the same comment twice

### ISSUE THREAD

**Issue by teofr** [rust-lang/rust-clippy#15755]
### Summary

When having a `SAFETY` comment on top of a statement that has an `unsafe` block, the statement is needed for the block, but is also linted for the (safe) statement.

I'm not entirely sure the best way to fix it easily, but I think a bigger refactor around these lints are needed, where: 
 - first `SAFETY` comments are gathered,
 - then each unsafe block is linked to a given SAFETY comment,
 - and finally lints are emitted accordingly

### Reproducer

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=07e7366f10d4bd79f6257b50dce7f9a5

I tried this code:

```rust
#![warn(clippy::undocumented_unsafe_blocks, clippy::unnecessary_safety_comment)]
#![no_main]

use std::num::NonZeroI32;

struct Date {
    value: NonZeroI32,
}

impl Date {
    const unsafe fn __from_ordinal_date_unchecked(year: i32, ordinal: u16) -> Self {
        Self {
            // Safety: The caller must guarantee that `ordinal` is not zero.
            value: unsafe { NonZeroI32::new_unchecked((year << 9) | ordinal as i32) },
        }
    }

    const fn into_julian_day_just_make_this_line_longer(self) -> i32 {
        42
    }
}

unsafe fn foo() {
    /// The Julian day of the Unix epoch.
    // SAFETY: fail ONLY if `accept-comment-above-attribute = false`
    const UNIX_EPOCH_JULIAN_DAY: i32 = unsafe { Date::__from_ordinal_date_unchecked(1970, 1) }
        .into_julian_day_just_make_this_line_longer();

    /// The Julian day of the Unix epoch.
    const UNIX_EPOCH_JULIAN_DAY_2: i32 = unsafe { Date::__from_ordinal_date_unchecked(1970, 1) }
        .into_julian_day_just_make_this_line_longer();

}
```

I expected to see this happen:

No lints around `UNIX_EPOCH_JULIAN_DAY`

Instead, this happened:

```
warning: constant has unnecessary safety comment
  --> src/lib.rs:26:5
   |
26 |     const UNIX_EPOCH_JULIAN_DAY: i32 = unsafe { Date::__from_ordinal_date_unchecked(1970, 1) }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: consider removing the safety comment
```

### Version

```text
rustc 1.92.0-nightly (4645a7988 2025-09-17)
binary: rustc
commit-hash: 4645a7988177c286f61609cc667ecae4c571a2e8
commit-date: 2025-09-17
host: aarch64-apple-darwin
release: 1.92.0-nightly
LLVM version: 21.1.1
```

### Additional Labels

@rustbot label +I-false-positive

**Comment by zihan0822** [comment]
Current impl does not consider `UNIX_EPOCH_JULIAN_DAY` as unsafe because there is a method call outside the unsafe block. It won't lint if you put the safety comment above the unsafe block or move the method call inside the unsafe block. 

Check this:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=44a514d26892ede289f1bf9cc38757a8
